libminifi/src/io/ZlibStream.cpp

```diff
@@ -0,0 +1,202 @@
+   * close the compressed stream.
+   */
+  do {
+    logger_->log_trace("writeData has %u B of input data left", strm_.avail_in);
+
+    strm_.next_out = outputBuffer_.data();
+    strm_.avail_out = outputBuffer_.size();
+
+    int flush = value == nullptr ? Z_FINISH : Z_NO_FLUSH;
+    logger_->log_trace("calling deflate with flush %d", flush);
+
+    int ret = deflate(&strm_, flush);
+    if (ret == Z_STREAM_ERROR) {
+      logger_->log_error("deflate failed, error code: %d", ret);
+      state_ = ZlibStreamState::ERRORED;
+      return -1;
+    }
+    int output_size = outputBuffer_.size() - strm_.avail_out;
+    logger_->log_trace("deflate produced %d B of output data", output_size);
+    if (BaseStream::writeData(outputBuffer_.data(), output_size) != output_size) {
+      logger_->log_error("Failed to write to underlying stream");
+      state_ = ZlibStreamState::ERRORED;
+      return -1;
+    }
+  } while (strm_.avail_out == 0);
+
+  return size;
+}
+
+void ZlibCompressStream::closeStream() {
+  if (state_ == ZlibStreamState::INITIALIZED) {
+    if (writeData(nullptr, 0U) == 0) {
+      state_ = ZlibStreamState::FINISHED;
+    }
+  }
+}
+
+/* ZlibDecompressStream */
+
+ZlibDecompressStream::ZlibDecompressStream(bool gzip)
+  : ZlibDecompressStream(this, gzip) {
+}
+
+ZlibDecompressStream::ZlibDecompressStream(DataStream* other, bool gzip)
+    : ZlibBaseStream(other)
+    , logger_(logging::LoggerFactory<ZlibDecompressStream>::getLogger()) {
+  int ret = inflateInit2(&strm_, 15 + (gzip ? 16 : 0) /* windowBits */);
+  if (ret != Z_OK) {
+    logger_->log_error("Failed to initialize z_stream with inflateInit2, error code: %d", ret);
+    throw Exception(ExceptionType::GENERAL_EXCEPTION, "zlib inflateInit2 failed");
+  }
+
+  state_ = ZlibStreamState::INITIALIZED;
+}
+
+ZlibDecompressStream::~ZlibDecompressStream() {
+  if (state_ != ZlibStreamState::UNINITIALIZED) {
+    inflateEnd(&strm_);
+  }
+}
+
+int ZlibDecompressStream::writeData(uint8_t* value, int size) {
+  if (state_ != ZlibStreamState::INITIALIZED) {
+    logger_->log_error("writeData called in invalid ZlibDecompressStream state, state is %hhu", state_);
+    return -1;
+  }
+
+  strm_.next_in = value;
+  strm_.avail_in = size;
+
+  /*
+   * inflate works similarly to deflate in that it will not leave input data unconsumed, and we have to watch avail_out,
+   * but in this case we do not have to close the stream, because it will detect the end of the compressed format
+   * and signal that it is ended by returning Z_STREAM_END and not accepting any more input data.
+   */
+  int ret;
+  do {
+    logger_->log_trace("writeData has %u B of input data left", strm_.avail_in);
+
+    strm_.next_out = outputBuffer_.data();
+    strm_.avail_out = outputBuffer_.size();
+
+    ret = inflate(&strm_, Z_NO_FLUSH);
+    if (ret == Z_STREAM_ERROR ||
+        ret == Z_NEED_DICT ||
+        ret == Z_DATA_ERROR ||
```

Comment:
Bleh, is this API designed so badly?

I would expect a success case or something, but handling some errors that look like a specific set of errors is weird.

Reply:
This is how it is designed, see the example usage: https://zlib.net/zlib_how.html